localstack/services/sns/publisher.py

```diff
@@ -0,0 +1,950 @@
+    return f"{external_url}/?Action=ConfirmSubscription&TopicArn={topic_arn}&Token={subscription_token}"
+
+
+def create_unsubscribe_url(external_url, subscription_arn):
+    return f"{external_url}/?Action=Unsubscribe&SubscriptionArn={subscription_arn}"
+
+
+class SubscriptionFilter:
+    def check_filter_policy_on_message_attributes(self, filter_policy, message_attributes):
+        if not filter_policy:
+            return True
+
+        for criteria in filter_policy:
+            conditions = filter_policy.get(criteria)
+            attribute = message_attributes.get(criteria)
+
+            if not self._evaluate_filter_policy_conditions(
+                conditions, attribute, message_attributes, criteria
+            ):
+                return False
+
+        return True
+
+    def _evaluate_filter_policy_conditions(
+        self, conditions, attribute, message_attributes, criteria
+    ):
+        if type(conditions) is not list:
+            conditions = [conditions]
+
+        tpe = attribute.get("DataType") or attribute.get("Type") if attribute else None
+        val = attribute.get("StringValue") or attribute.get("Value") if attribute else None
+        if attribute is not None and tpe == "String.Array":
+            values = ast.literal_eval(val)
+            for value in values:
+                for condition in conditions:
+                    if self._evaluate_condition(value, condition, message_attributes, criteria):
+                        return True
+        else:
+            for condition in conditions:
+                value = val or None
+                if self._evaluate_condition(value, condition, message_attributes, criteria):
+                    return True
+
+        return False
+
+    def _evaluate_condition(self, value, condition, message_attributes, criteria):
+        if type(condition) is not dict:
+            return value == condition
+        elif condition.get("exists") is not None:
+            return self._evaluate_exists_condition(
+                condition.get("exists"), message_attributes, criteria
+            )
+        elif value is None:
+            # the remaining conditions require the value to not be None
+            return False
+        elif condition.get("anything-but"):
+            return value not in condition.get("anything-but")
+        elif condition.get("prefix"):
+            prefix = condition.get("prefix")
+            return value.startswith(prefix)
+        elif condition.get("numeric"):
+            return self._evaluate_numeric_condition(condition.get("numeric"), value)
+        return False
+
+    @staticmethod
+    def _is_number(x):
+        try:
+            float(x)
+            return True
+        except ValueError:
+            return False
+
+    def _evaluate_numeric_condition(self, conditions, value):
+        if not self._is_number(value):
+            return False
+
+        for i in range(0, len(conditions), 2):
+            value = float(value)
```

Comment:
nit: this could be outside the loop. Optimize all the things!

Reply:
Perfect, thanks! I've removed `_is_number` because it was casting to `float` as well and I'm doing it once at the beginning of the function. Thanks!